README.md
# BEGIN::Lift

Lift subroutine calls into the BEGIN phase.

# Copyright and License

This software is copyright (c) 2016 by Stevan Little.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Perl extension BEGIN::Lift

0.07
    - remove the B::CompilerPhase::Hook dependency
      and replace it with Devel::Hook

0.06 2017-08-19
    - managed to upload the last one without a
      Makefile.PL, this should fix this.

0.05 2017-08-14
    - remove enforcement of BEGIN time, it will
      just work at runtime as well
        - removed test that tested this

0.04 2017-01-13
    - better now that the synopsis actually works

0.03 2017-01-13
    - actually include a Makefile.PL in the release ;)

0.02 2016-12-29
	- adding some missing POD, this is what I
	  get for not using Dist::Zilla on this one

0.01 2016-12-29
    - Initial release to an unsuspecting world


lib/BEGIN/Lift.pm
package BEGIN::Lift;
# ABSTRACT: Lift subroutine calls into the BEGIN phase

use strict;
use warnings;

our $VERSION;
our $AUTHORITY;

use Sub::Name              ();
use B::CompilerPhase::Hook ();

use Devel::CallParser;
use XSLoader;
BEGIN {
    $VERSION   = '0.04';
    $AUTHORITY = 'cpan:STEVAN';
    XSLoader::load( __PACKAGE__, $VERSION );
}

sub install {
    my ($pkg, $method, $handler) = @_;

    # need to force a new CV each time here
    # not entirely sure why, but I assume
    # that perl was trying to optimize things
    # which is not what I actually want.
    my $cv = eval 'sub {}';

    # now we need to install the stub
    # we just created, but first we need to
    # verify that we are the only ones using
    # the typeglob we are installing into.
    # This makes it easier/safer to delete
    # the stub before runtime.
    {
        no strict 'refs';
        die "Cannot install the lifted keyword ($method) into package ($pkg) when that typeglob (\*${pkg}::${method}) already exists"
            if exists ${"${pkg}::"}{$method};
        *{"${pkg}::${method}"} = $cv;
    }

    # give the handler a name so that
    # it shows up sensibly in stack
    # traces and the like ...
    Sub::Name::subname( "${pkg}::${method}", $handler );

    # install the keyword handler ...
    BEGIN::Lift::Util::install_keyword_handler(
        $cv, sub { $handler->( $_[0] ? $_[0]->() : () ) }
    );

    # clean things up ...
    B::CompilerPhase::Hook::enqueue_UNITCHECK {
        no strict 'refs';
        # NOTE:
        # this is safe only because we
        # confirmed above that there was
        # no other use of this typeglob
        # and so it is ok to delete
        delete ${"${pkg}::"}{$method}
    };
}

1;

__END__

=pod

=head1 NAME

BEGIN::Lift - Lift subroutine calls into the BEGIN phase

=head1 SYNOPSIS

  package Cariboo;
  use strict;
  use warnings;

  use BEGIN::Lift;

  sub import {
      my $caller = caller;

      BEGIN::Lift::install(
          ($caller, 'extends') => sub {
              no strict 'refs';
              @{$caller . '::ISA'} = @_;
          }
      );
  }

  package Foo;
  use Cariboo;

  extends 'Bar';

  # functionally equivalent to ...
  # BEGIN { @ISA = ('Bar') }

=head1 DESCRIPTION

This module serves a very specific purpose, which is to provide a
mechanism through which we can "lift" a given subroutine to be
executed entirely within the C<BEGIN> phase of the Perl compiler
and to leave no trace of itself in the C<RUN> phase.

=head2 Modules loaded at runtime?

If a package that uses this module is loaded at runtime (perhaps
via the C<require> builtin), it will still work correctly (to the
best of my knowledge that is).

=head1 FUNCTIONS

=head2 C<install( $package, $keyword_name, $keyword_handler )>

This will install a lifted subroutine named C<$keyword_name> into
the specified C<$package>. All calls to this the lifted subroutine
will execute the C<$keyword_handler> immediately after parsing it.

If this subroutine is called outside of the C<BEGIN> phase, an error
will be thrown. If there already exists a typeglob for C<$keyword_name>
then an error will be thrown.

=head1 CAVEATS

=head2 All arguments to lifted subroutines must be C<BEGIN> time safe

This means they require no runtime initiatlization or access to
runtime initialized variables (as they won't be initialized).

For instance, given a lifted subroutine called C<add>, this code is
C<BEGIN> time safe because the arguments are numeric literals.

  add( 1, 1 );

While this version is not safe because it relies on the C<@args>
variable being initialized at runtime.

  my @args = (1, 1);
  add( @args );

=head1 PARSING ISSUES

Ideally we can (eventually) detect these situations and error
accordingly so that this is no longer a burden to the user of this
module, but instead just part of the normal operation of it.

=head2 Non-void context

If, for instance, a lifted sub is called such that the return value
is to be assigned to a variable, such as:

  my $x = my_lifted_sub();

It will not behave as expected, since C<my_lifted_sub> is evaluated
entirely at C<BEGIN> time, the resulting value for C<$x> at C<RUN>
time is C<undef>.

=head2 Expression context

If, for instance, a lifted sub is called within an expression where
the return value is important, such as:

  if ( my_lifted_sub() && 10 ) { ... }

It will not behave as expected, since C<my_lifted_sub> is evaluated
entirely at C<BEGIN> time and has the value of C<undef> at runtime,
the conditional will always fail.

=head2 Statement modifier context

If, for instance, a lifted sub call is guarded by a statement modifier,
such as:

  my_lifted_sub() if 0;

It will not behave as expected, since the lifted sub call is evaluated
entirely at C<BEGIN> time the statement modifier has no affect at all
and <my_lifted_sub> will always be executed.

=head1 SEE ALSO

=over 4

=item L<Devel::BeginLift>

This does a similar thing, but does it via "some slightly insane perlguts magic",
while this module has much the same goals, it will (hopefully) accomplish it with
less insanity.

=back

=head1 AUTHOR

Stevan Little <stevan@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Stevan Little.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut




lib/BEGIN/KeywordLift.pm
package BEGIN::KeywordLift;

use strict;
use warnings;

our $VERSION;
our $AUTHORITY;

use Sub::Name ();

use Devel::CallParser;
use XSLoader;
BEGIN {
    $VERSION   = '0.01';
    $AUTHORITY = 'cpan:STEVAN';
    XSLoader::load( __PACKAGE__, $VERSION );
}

sub install {
    my ($pkg, $method, $handler) = @_;

    die 'Lifted keywords must be created during BEGIN time, not (' . ${^GLOBAL_PHASE}. ')'
        unless ${^GLOBAL_PHASE} eq 'START';

    # need to force a new CV each time here
    my $cv = eval 'sub {}';

    {
        no strict 'refs';

        die "Cannot install the lifted keyword ($method) into package ($pkg) when that typeglob (\*${pkg}::${method}) already exists"
            if exists ${"${pkg}::"}{$method};

        *{"${pkg}::${method}"} = $cv;
    }

    Sub::Name::subname( "${pkg}::${method}", $handler );

    BEGIN::KeywordLift::Util::install_keyword_handler(
        $cv,
        sub {
            # read till the end of the statement ...
            my $stmt = BEGIN::KeywordLift::Util::parse_full_statement;
            # then execute that callback and pass the
            # result to the handler, this basically
            # evaluates all the arguments, so make
            # sure they are BEGIN time clean
            $handler->( $stmt->() );
            return (sub {}, 1);
        }
    );

    BEGIN::KeywordLift::Util::install_keyword_cleanup_handler(
        sub {
            no strict 'refs';
            delete ${"${pkg}::"}{$method}
        }
    );
}

1;

__END__

=pod

=head1 NAME

BEGIN::KeywordLift - Lift subroutine calls into the BEGIN phase

=head1 SYNOPSIS

    package Cariboo;
    use strict;
    use warnings;

    use BEGIN::KeywordLift;

    sub import {
        my $caller = caller;

        BEGIN::KeywordLift::install(
            ($caller, 'extends') => sub {
                no strict 'refs';
                @{$caller . '::ISA'} = @_;
            }
        );
    }

    package Foo;
    use Cariboo;

    extends 'Bar';

    # functionally equivalent to ...
    # BEGIN { @ISA = ('Bar') }

=head1 DESCRIPTION

=cut




lib/Keyword/BeginLift.pm
package Keyword::BeginLift;

use strict;
use warnings;

our $VERSION;
our $AUTHORITY;

use Sub::Name ();

use Devel::CallParser;
use XSLoader;
BEGIN {
    $VERSION   = '0.01';
    $AUTHORITY = 'cpan:STEVAN';
    XSLoader::load( __PACKAGE__, $VERSION );
}

sub install {
    my ($pkg, $method, $handler) = @_;

    die 'Lifted keywords must be created during BEGIN time, not (' . ${^GLOBAL_PHASE}. ')'
        unless ${^GLOBAL_PHASE} eq 'START';

    # need to force a new CV each time here
    my $cv = eval 'sub {}';

    {
        no strict 'refs';

        die "Cannot install the lifted keyword ($method) into package ($pkg) when that typeglob (\*${pkg}::${method}) already exists"
            if exists ${"${pkg}::"}{$method};

        *{"${pkg}::${method}"} = $cv;
    }

    Sub::Name::subname( "${pkg}::${method}", $handler );

    Keyword::BeginLift::Util::install_keyword_handler(
        $cv,
        sub {
            # read till the end of the statement ...
            my $stmt = Keyword::BeginLift::Util::parse_full_statement;
            # then execute that callback and pass the
            # result to the handler, this basically
            # evaluates all the arguments, so make
            # sure they are BEGIN time clean
            $handler->( $stmt->() );
            return (sub {}, 1);
        }
    );

    Keyword::BeginLift::Util::install_keyword_cleanup_handler(
        sub {
            no strict 'refs';
            delete ${"${pkg}::"}{$method}
        }
    );
}

1;

__END__

=pod

=head1 NAME

Keyword::BeginLift - Lift subroutine calls into the BEGIN phase

=head1 SYNOPSIS

    package Cariboo;
    use strict;
    use warnings;

    use Keyword::BeginLift;

    sub import {
        my $caller = caller;

        Keyword::BeginLift::install(
            ($caller, 'extends') => sub {
                no strict 'refs';
                @{$caller . '::ISA'} = @_;
            }
        );
    }

    package Foo;
    use Cariboo;

    extends 'Bar';

    # functionally equivalent to ...
    # BEGIN { @ISA = ('Bar') }

=head1 DESCRIPTION

=cut




Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker 6.30;

my %WriteMakefileArgs = (
    'NAME'               => 'BEGIN::Lift',
    'VERSION'            => '0.05',
    'DISTNAME'           => 'BEGIN-Lift',
    'ABSTRACT'           => 'Lift subroutine calls into the BEGIN phase',
    'AUTHOR'             => 'Stevan Little',
    'BUILD_REQUIRES'     => {},
    'CONFIGURE_REQUIRES' => {
        'Devel::CallParser'   => '0',
        'ExtUtils::MakeMaker' => '6.30'
    },
    'EXE_FILES'          => [],
    'LICENSE'            => 'perl',
    'NORECURS'           => 1,
    'PREREQ_PM'  => {
        # cpan ....
        'Sub::Name'              => 0,
        'B::CompilerPhase::Hook' => '0.04',
        'Devel::CallParser'      => 0,
        'XSLoader'               => 0,
        # tests ..
        'Test::More'             => 0,
    },
    'clean'              => { FILES => 'cover_db/ nytprof/ callparser1.h' },
    'test'               => { RECURSIVE_TEST_FILES => 1 },
);

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

use File::Spec::Functions 'abs2rel';
use Devel::CallParser 'callparser1_h', 'callparser_linkable';
open my $fh, '>', 'callparser1.h' or die "Couldn't write to callparser1.h";
$fh->print(callparser1_h);
my @linkable = map { abs2rel($_) } callparser_linkable;
unshift @linkable, '$(BASEEXT)$(OBJ_EXT)' if @linkable;
$WriteMakefileArgs{OBJECT} = join(' ', @linkable) if @linkable;

WriteMakefile(%WriteMakefileArgs);
